boj/silver/나무_탈출/jaeung.js

```diff
@@ -0,0 +1,47 @@
+const fs = require("fs");
+const input = fs.readFileSync("dev/stdin").toString().trim().split("\n");
+
+function solution(input) {
+  const N = +input[0];
+  const visited = Array(N + 1).fill(false);
+  const nodes = Array.from(Array(N + 1), () => []);
+  const stack = [[1, 0]];
+  let moveCount = 0;
+
+  visited[1] = true;
+
+  for (let i = 1; i < N; i++) {
+    const line = input[i].split(" ");
+    const a = +line[0];
+    const b = +line[1];
+
+    nodes[a].push(b);
+    nodes[b].push(a);
+  }
+
+  loop: while (stack.length > 0) {
+    const [peek, depth] = stack[stack.length - 1];
+    const isLeafNode = nodes[peek].length === 1;
+
+    for (const node of nodes[peek]) {
+      const isVisit = visited[node] === true;
+
+      if (!isVisit) {
+        stack.push([node, depth + 1]);
+        visited[node] = true;
+        continue loop;
+      }
+    }
+
+    if (isLeafNode) {
+      moveCount += depth;
+    }
+
+    stack.pop();
+  }
```

Comment:
저랑 같은 접근 방식인데 풀이가 다르군요!
로직을 아래 코드처럼 변경하면 코드의 흐름을 파악하기 더 쉬울 것 같습니다~
```suggestion
  while (stack.length > 0) {
    const [peek, depth] = stack.pop();
    const isLeafNode = nodes[peek].length === 1;

    if (isLeafNode) {
      moveCount += depth;
      continue;
    }

    for (const node of nodes[peek]) {
      const isVisit = visited[node] === true;

      if (!isVisit) {
        stack.push([node, depth + 1]);
        visited[node] = true;
      }
    }
  }
```

Reply:
제안해주신 내용을 보고 발상을 전환해서 문제를 해결할 수 있었네요! 😁

제가 기존 코드를 저렇게 작성한 이유는, **노드 방문 시 마다 `stack` 배열을 통해 노드가 어떤 노드를 거쳐서 왔는지 확인할 수 있고, 순서를 고려할 수 있다**는 점 때문이었습니다.

다만, **깊이가 깊어질수록 단순히 `pop`을 하는 과정 때문에 시간 복잡도가 늘어난다**는 것인데, 말씀해주신 내용대로라면 이 과정이 생략되어 시간 복잡도를 더 줄일 수 있었습니다.

덕분에 문제에 필요 없는 기능(노드의 방문 경로)을 덜고, 필요한 기능(리프 노드의 깊이를 확인하는 것)만 넣어서 정답 처리가 되었네요!

좋은 리뷰 감사드립니다. 🙇‍♂️